In [4]:
# Import libarires
import pandas as pd
import numpy as np
import os

In [5]:
# Create path
path = r'C:\Users\SuperHome\Documents\Instacart Basket Analysis\02 Data'

In [6]:
# Import products dataframe
df_pro = pd.read_csv(os.path.join(path,'Original Data','products.csv'),index_col = False)

In [7]:
# products describe
df_pro.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

# Conclusion before checking data consistnecy
As shows above, this dataframe contains 49693 rows but only 49688 product_id, as the product_id is unique for 
every proudtc, there might be duplicates on this dataframe.
The prices column has a maximum price 99999 which is way higher than the other price, so it would be an outlier.
Except those two, the others look all good so far.

In [8]:
#Task1 - Data consistency Checks -- products dataframe

In [9]:
# 01 Checking mixed-type data
for col in df_pro.columns.tolist():
    weird = (df_pro[[col]].applymap(type) != df_pro[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_pro[weird]) > 0:
        print(col)

product_name


As showing the product_name column contains mixed type data, as the name could be object(string)type, there
might be any empty values in it. Will check again after the whole process.

In [10]:
# 02 check Missing values
df_pro.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

# The product_name contains 16 nan values.

In [11]:
# Finding those missing values
df_pro[df_pro['product_name'].isnull() == True]

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [12]:
# Creating new dataframe without deleting those nan values
df_pro_clean = df_pro[df_pro['product_name']. isnull() == False]

In [13]:
df_pro_clean.shape # 16 rows were put away in this new dataframe.

(49677, 5)

In [14]:
# Checking mixed-type data for this new dataframe
for col in df_pro_clean.columns.tolist():
    weird = (df_pro_clean[[col]].applymap(type) != df_pro_clean[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_pro_clean[weird]) > 0:
        print(col) # no showing means we ignor the nan values and the product_name now has the same data type.

In [15]:
# Finding duplicates
df_pro_clean_dp = df_pro_clean[df_pro_clean.duplicated()]

In [16]:
df_pro_clean_dp # There are 5 duplicates found which is macth the guessing that the product_id is 49688 in total,
                 # but the whole dataframe has 5 more rows than it.

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [17]:
# Delating the duplicates
df_pro_all_clean = df_pro_clean.drop_duplicates()

In [18]:
df_pro_all_clean.shape # 5 duplicates were delated 

(49672, 5)

In [22]:
# Data Consistency Check -- orders dataframe

In [20]:
# Import orders
df_ord = pd.read_csv(os.path.join(path,'Prepared Data','orders_wrangled.csv'),index_col = False)

In [21]:
# Orders dataframe describe
df_ord.describe() # Why there apears an extra "Unname Column"

Unnamed: 0      order_id       user_id  order_number   day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [22]:
# Task 3 Check for mixed-type data in your df_ords dataframe.
for col in df_ord.columns.tolist():
    weird = (df_ord[[col]].applymap(type) != df_ord[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_ord[weird]) > 0:
        print(col) 

# Task 4 If you find mixed-type data, fix it.
There are no columns with mixed-type data in this dataframe.

In [24]:
# Task 5 Run a check for missing values in your df_ords dataframe
df_ord.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
day_of_week                    0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [ ]:
There are 206209 rows with missing values in 'days_since_prior_order' column, as this is a big number, it might
be caused by human error or any no on time updating.

In [25]:
# Task 6 Address the missing values using an appropriate method 
df_ord['days_since_prior_order'].value_counts(dropna=False, normalize=True)

30.0    0.107955
7.0     0.093715
6.0     0.070157
4.0     0.064803
3.0     0.063432
5.0     0.062700
NaN     0.060276
2.0     0.056475
8.0     0.053117
1.0     0.042456
9.0     0.034547
14.0    0.029298
10.0    0.027823
13.0    0.024324
11.0    0.023668
12.0    0.022258
0.0     0.019805
15.0    0.019461
16.0    0.013721
21.0    0.013291
17.0    0.011472
20.0    0.011262
18.0    0.010488
19.0    0.010051
22.0    0.009357
28.0    0.007827
23.0    0.006982
27.0    0.006435
24.0    0.006054
25.0    0.005622
29.0    0.005610
26.0    0.005558
Name: days_since_prior_order, dtype: float64

##### As showing above we have 6% of the nan values in 'days_since_prior_order' which is a quite big percentage.
We can not delete those nan values and it will cause the dataframe skewed, in this case we have the mean 11 and 
the median is 7, because this is not the normal distribution as the mean is much higher than the median, so 
in this case I will choose fill out the nan values with median 7 insetad of the other ways.

In [26]:
# Addressing the nan values with median
df_ord['days_since_prior_order'].fillna(7,inplace = True)

In [27]:
df_ord.describe() # we filled out the nan values with median 7 and the mean is smaller than original.

Unnamed: 0      order_id       user_id  order_number   day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.421083e+06  
mean        1.345202e+01            1.086681e+01  
std         4.226088e+00            8.978521e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            5.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [28]:
# Task 7 Run a check for duplicate values in your df_ords data.
df_ord_dp = df_ord[df_ord.duplicated()]

In [29]:
df_ord_dp # There is no duplicates found in this dataframe as showing below.

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

# Task 8 Address the duplicates using an appropriate method.
No duplicates found.

In [ ]:
# Task 9 Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder
 # and give them appropriate, succinct names.

In [32]:
# Export cleaned proudcts
df_pro_all_clean.to_csv(os.path.join(path,'Prepared Data','product_clean.csv'))

In [31]:
# Export cleaned orders
df_ord.to_csv(os.path.join(path,'Prepared Data','orders_clean.csv'))